In [72]:
import pandas as pd #tablas https://pandas.pydata.org/docs/user_guide/pyarrow.html
import numpy as np #estadistica https://numpy.org/doc/1.25/user/index.html#user
import matplotlib.pyplot as plt #gráficos https://matplotlib.org/stable/plot_types/index.html
import seaborn as sns #gráficos https://seaborn.pydata.org/tutorial/regression.html
from sklearn import linear_model #modelo de regresion https://scikit-learn.org/stable/
from sklearn.preprocessing import MinMaxScaler # do not use the function Normalise() - it does something entirely different
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from datetime import datetime #https://docs.python.org/3/library/datetime.html
from datetime import timedelta #this is a good time to learn about dates and timestamps. You can find some info here:
from sklearn.linear_model import LogisticRegression
from sklearn.utils import resample
from imblearn.over_sampling import SMOTE
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.neighbors import KNeighborsClassifier
from scipy.stats import ttest_1samp
import scipy.stats as st
from sklearn.neighbors import KNeighborsClassifier
from mlxtend.plotting import plot_decision_regions
from datetime import datetime, timedelta
from sqlalchemy import create_engine  # for establishing the connection and authentication
from getpass import getpass  # To get the password without showing the input
import pandas_market_calendars as mcal

1. https://developers.refinitiv.com/en/article-catalog/article/finding-expired-options-and-backtesting-a-short-iron-condor-stra

As there is no free data available; following the article I'm taking the following decissions:
    - Expiration date = third Friday of each month
    - Striking price = strike price of 15th September 2023
    - Applying a linear reggression analysis we will determine what amount of volume each option has per period. 
    - Volume per period = BATS daily 01/01/2013 until 25/05/23 (assumption: all transactions of options in that period for the ones that expire in the period)

1. Expiration date (third Friday of each month) WE NEED TO CHANGE THE ONES THAT ARE HOLIDAYS TO T-1

In [73]:
def get_third_fridays(start_date, end_date):
    third_fridays = []

    current_date = start_date
    while current_date <= end_date:
        if current_date.weekday() == 4: 
            day_of_month = current_date.day
            if 15 <= day_of_month <= 21:
                third_fridays.append(current_date)

        current_date += timedelta(days=1)

    return third_fridays

def adjust_to_previous_business_day(date, calendar):
    while date.weekday() >= 5 or date in calendar.holidays().holidays:
        date -= timedelta(days=1)
    return date

start_date = datetime(2013, 1, 1)
end_date = datetime(2023, 8, 25)

third_fridays_list = get_third_fridays(start_date, end_date)

CBOE = mcal.get_calendar('CBOE_Index_Options')

adjusted_third_fridays = [adjust_to_previous_business_day(date, CBOE) for date in third_fridays_list]

expr_dt = pd.DataFrame(adjusted_third_fridays, columns=['expiration_date'])
expr_dt.describe()


/var/folders/dm/hhj01rls7tz3g5v4m3hhj1bw0000gn/T/ipykernel_82331/3383185892.py:30: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  expr_dt.describe()


expiration_date
count                   128
unique                  128
top     2013-01-18 00:00:00
freq                      1
first   2013-01-18 00:00:00
last    2023-08-18 00:00:00

In [107]:
expr_dt

expiration_date
0        2013-01-18
1        2013-02-15
2        2013-03-15
3        2013-04-19
4        2013-05-17
..              ...
123      2023-04-21
124      2023-05-19
125      2023-06-16
126      2023-07-21
127      2023-08-18

[128 rows x 1 columns]

2. Let's bring todays information option

In [75]:
password = getpass()

In [76]:
connection_string = 'mysql+pymysql://root:'+password+'@localhost/raw_data'
engine = create_engine(connection_string)
query = '''select c.company_id FROM companies as c;'''

companies = pd.read_sql_query(query, engine)

In [77]:
companies

company_id
0       TSLA
1       META
2       GOOG
3        NIO
4       AAPL
5       AMZN
6       MSFT

In [78]:
companies["company_id"].unique()

array(['TSLA', 'META', 'GOOG', 'NIO', 'AAPL', 'AMZN', 'MSFT'],
      dtype=object)

https://pypi.org/project/yfinance/

In [79]:

import yfinance as yf

In [80]:
msft = yf.Ticker("MSFT")
opt = msft.option_chain('2023-09-15') #option chain from 15th october 2023
opt

Options(calls=         contractSymbol             lastTradeDate  strike  lastPrice    bid  \
0   MSFT230915C00110000 2023-08-23 17:43:20+00:00   110.0     219.17    0.0   
1   MSFT230915C00115000 2023-08-10 19:08:16+00:00   115.0     208.45    0.0   
2   MSFT230915C00120000 2023-08-18 14:08:27+00:00   120.0     192.91    0.0   
3   MSFT230915C00125000 2023-08-24 19:29:04+00:00   125.0     197.25    0.0   
4   MSFT230915C00130000 2023-07-14 17:23:13+00:00   130.0     217.00  190.0   
..                  ...                       ...     ...        ...    ...   
94  MSFT230915C00520000 2023-08-23 13:39:29+00:00   520.0       0.01    0.0   
95  MSFT230915C00525000 2023-08-07 14:30:04+00:00   525.0       0.01    0.0   
96  MSFT230915C00530000 2023-08-07 14:21:44+00:00   530.0       0.01    0.0   
97  MSFT230915C00535000 2023-07-25 19:57:55+00:00   535.0       0.13    0.0   
98  MSFT230915C00540000 2023-08-01 17:25:41+00:00   540.0       0.02    0.0   

       ask  change  percentChange  vo

In [81]:
calls_df = opt.calls
puts_df = opt.puts

In [82]:
calls_df

contractSymbol             lastTradeDate  strike  lastPrice    bid  \
0   MSFT230915C00110000 2023-08-23 17:43:20+00:00   110.0     219.17    0.0   
1   MSFT230915C00115000 2023-08-10 19:08:16+00:00   115.0     208.45    0.0   
2   MSFT230915C00120000 2023-08-18 14:08:27+00:00   120.0     192.91    0.0   
3   MSFT230915C00125000 2023-08-24 19:29:04+00:00   125.0     197.25    0.0   
4   MSFT230915C00130000 2023-07-14 17:23:13+00:00   130.0     217.00  190.0   
..                  ...                       ...     ...        ...    ...   
94  MSFT230915C00520000 2023-08-23 13:39:29+00:00   520.0       0.01    0.0   
95  MSFT230915C00525000 2023-08-07 14:30:04+00:00   525.0       0.01    0.0   
96  MSFT230915C00530000 2023-08-07 14:21:44+00:00   530.0       0.01    0.0   
97  MSFT230915C00535000 2023-07-25 19:57:55+00:00   535.0       0.13    0.0   
98  MSFT230915C00540000 2023-08-01 17:25:41+00:00   540.0       0.02    0.0   

       ask  change  percentChange  volume  openInterest  impliedVolatility  \
0     0.00     0.0            0.0       1             0           0.000010   
1     0.00     0.0            0.0       1             0           0.000010   
2     0.00     0.0            0.0       3             0           0.000010   
3     0.00     0.0            0.0      20             0           0.000010   
4   192.95     0.0            0.0       1           382           0.000010   
..     ...     ...            ...     ...           ...                ...   
94    0.00     0.0            0.0       9             0           0.500005   
95    0.00     0.0            0.0       3             0           0.500005   
96    0.00     0.0            0.0       3             0           0.500005   
97    2.13     0.0            0.0       3             8           1.166020   
98    0.00     0.0            0.0      15             0           0.500005   

    inTheMoney contractSize currency  
0         True      REGULAR      USD  
1         True      REGULAR      USD  
2         True      REGULAR      USD  
3         True      REGULAR      USD  
4         True      REGULAR      USD  
..         ...          ...      ...  
94       False      REGULAR      USD  
95       False      REGULAR      USD  
96       False      REGULAR      USD  
97       False      REGULAR      USD  
98       False      REGULAR      USD  

[99 rows x 14 columns]

In [83]:
puts_df

contractSymbol             lastTradeDate  strike  lastPrice    bid  \
0   MSFT230915P00110000 2023-08-11 18:31:24+00:00   110.0       0.01    0.0   
1   MSFT230915P00115000 2023-08-16 17:13:48+00:00   115.0       0.01    0.0   
2   MSFT230915P00120000 2023-07-26 17:53:51+00:00   120.0       0.01    0.0   
3   MSFT230915P00125000 2023-07-20 16:22:01+00:00   125.0       0.01    0.0   
4   MSFT230915P00130000 2023-07-27 18:58:26+00:00   130.0       0.02    0.0   
..                  ...                       ...     ...        ...    ...   
89  MSFT230915P00495000 2023-08-23 14:06:36+00:00   495.0     168.32    0.0   
90  MSFT230915P00500000 2023-08-16 18:39:22+00:00   500.0     176.60    0.0   
91  MSFT230915P00505000 2023-08-16 18:39:22+00:00   505.0     181.50    0.0   
92  MSFT230915P00510000 2023-08-16 18:39:22+00:00   510.0     186.15    0.0   
93  MSFT230915P00520000 2023-07-25 15:15:29+00:00   520.0     172.75  195.6   

       ask  change  percentChange  volume  openInterest  impliedVolatility  \
0     0.00     0.0            0.0     1.0             0           0.500005   
1     0.00     0.0            0.0     1.0             0           0.500005   
2     0.02     0.0            0.0     1.0          2510           1.375003   
3     0.03     0.0            0.0     6.0          1153           1.375003   
4     0.05     0.0            0.0     3.0          3955           1.375003   
..     ...     ...            ...     ...           ...                ...   
89    0.00     0.0            0.0     4.0             0           0.000010   
90    0.00     0.0            0.0   145.0             0           0.000010   
91    0.00     0.0            0.0     NaN             0           0.000010   
92    0.00     0.0            0.0     NaN             0           0.000010   
93  196.45     0.0            0.0     NaN             0           0.851564   

    inTheMoney contractSize currency  
0        False      REGULAR      USD  
1        False      REGULAR      USD  
2        False      REGULAR      USD  
3        False      REGULAR      USD  
4        False      REGULAR      USD  
..         ...          ...      ...  
89        True      REGULAR      USD  
90        True      REGULAR      USD  
91        True      REGULAR      USD  
92        True      REGULAR      USD  
93        True      REGULAR      USD  

[94 rows x 14 columns]

FUNCTION TO BRING ALL COMPANIES CALLS AND PUTS INFORMATION ON THE 15TH SEPTEMBER 2023

In [84]:

def get_options_data(tickers, expiration_date):
    options_data = []

    for ticker in tickers:
        stock = yf.Ticker(ticker)
        opt = stock.option_chain(expiration_date)
        call_df = opt.calls
        put_df = opt.puts
        
        call_df['Type'] = 'Call'
        put_df['Type'] = 'Put'
        
        call_df['Ticker'] = ticker
        put_df['Ticker'] = ticker
        
        options_data.append(call_df)
        options_data.append(put_df)

    all_options_df = pd.concat(options_data, ignore_index=True)

    return all_options_df


tickers = companies["company_id"].unique() 


expiration_date = '2023-09-15'


options_df = get_options_data(tickers, expiration_date)

options_df

contractSymbol             lastTradeDate  strike  lastPrice    bid  \
0     TSLA230915C00001670 2023-08-24 16:38:35+00:00    1.67     231.66    0.0   
1     TSLA230915C00003330 2023-08-25 14:33:55+00:00    3.33     231.10    0.0   
2     TSLA230915C00005000 2023-08-17 14:49:44+00:00    5.00     220.15    0.0   
3     TSLA230915C00006670 2023-07-28 19:25:46+00:00    6.67     258.73  229.3   
4     TSLA230915C00008330 2023-07-28 19:25:46+00:00    8.33     257.08  227.9   
...                   ...                       ...     ...        ...    ...   
1992  MSFT230915P00495000 2023-08-23 14:06:36+00:00  495.00     168.32    0.0   
1993  MSFT230915P00500000 2023-08-16 18:39:22+00:00  500.00     176.60    0.0   
1994  MSFT230915P00505000 2023-08-16 18:39:22+00:00  505.00     181.50    0.0   
1995  MSFT230915P00510000 2023-08-16 18:39:22+00:00  510.00     186.15    0.0   
1996  MSFT230915P00520000 2023-07-25 15:15:29+00:00  520.00     172.75  195.6   

         ask  change  percentChange  volume  openInterest  impliedVolatility  \
0       0.00     0.0            0.0     3.0           0.0           0.000010   
1       0.00     0.0            0.0     5.0           0.0           0.000010   
2       0.00     0.0            0.0     2.0           0.0           0.000010   
3     233.75     0.0            0.0     3.0          29.0          10.593753   
4     232.20     0.0            0.0     1.0          17.0           9.832035   
...      ...     ...            ...     ...           ...                ...   
1992    0.00     0.0            0.0     4.0           0.0           0.000010   
1993    0.00     0.0            0.0   145.0           0.0           0.000010   
1994    0.00     0.0            0.0     NaN           0.0           0.000010   
1995    0.00     0.0            0.0     NaN           0.0           0.000010   
1996  196.45     0.0            0.0     NaN           0.0           0.851564   

      inTheMoney contractSize currency  Type Ticker  
0           True      REGULAR      USD  Call   TSLA  
1           True      REGULAR      USD  Call   TSLA  
2           True      REGULAR      USD  Call   TSLA  
3           True      REGULAR      USD  Call   TSLA  
4           True      REGULAR      USD  Call   TSLA  
...          ...          ...      ...   ...    ...  
1992        True      REGULAR      USD   Put   MSFT  
1993        True      REGULAR      USD   Put   MSFT  
1994        True      REGULAR      USD   Put   MSFT  
1995        True      REGULAR      USD   Put   MSFT  
1996        True      REGULAR      USD   Put   MSFT  

[1997 rows x 16 columns]

In [85]:
options_df.isna().sum()

contractSymbol        0
lastTradeDate         0
strike                0
lastPrice             0
bid                   2
ask                   2
change                0
percentChange         0
volume               58
openInterest          3
impliedVolatility     0
inTheMoney            0
contractSize          0
currency              0
Type                  0
Ticker                0
dtype: int64

We can assume that it's becuase there is no transaction there for = 0

In [86]:
options_15sep23= options_df.fillna(0)
options_15sep23.head(10)

contractSymbol             lastTradeDate  strike  lastPrice     bid  \
0  TSLA230915C00001670 2023-08-24 16:38:35+00:00    1.67     231.66    0.00   
1  TSLA230915C00003330 2023-08-25 14:33:55+00:00    3.33     231.10    0.00   
2  TSLA230915C00005000 2023-08-17 14:49:44+00:00    5.00     220.15    0.00   
3  TSLA230915C00006670 2023-07-28 19:25:46+00:00    6.67     258.73  229.30   
4  TSLA230915C00008330 2023-07-28 19:25:46+00:00    8.33     257.08  227.90   
5  TSLA230915C00010000 2023-07-03 14:38:24+00:00   10.00     270.79  243.65   
6  TSLA230915C00011670 2023-08-21 18:27:04+00:00   11.67     217.97    0.00   
7  TSLA230915C00013330 2023-07-28 19:25:46+00:00   13.33     252.08  223.10   
8  TSLA230915C00015000 2023-08-24 13:58:48+00:00   15.00     218.48    0.00   
9  TSLA230915C00016670 2023-07-25 13:58:23+00:00   16.67     254.06  211.95   

      ask  change  percentChange  volume  openInterest  impliedVolatility  \
0    0.00     0.0            0.0     3.0           0.0           0.000010   
1    0.00     0.0            0.0     5.0           0.0           0.000010   
2    0.00     0.0            0.0     2.0           0.0           0.000010   
3  233.75     0.0            0.0     3.0          29.0          10.593753   
4  232.20     0.0            0.0     1.0          17.0           9.832035   
5  246.50     0.0            0.0     1.0          39.0           0.000000   
6    0.00     0.0            0.0     1.0           0.0           0.000010   
7  227.30     0.0            0.0     2.0          91.0           8.177739   
8    0.00     0.0            0.0     1.0           0.0           0.000010   
9  214.15     0.0            0.0    49.0         133.0           0.000010   

   inTheMoney contractSize currency  Type Ticker  
0        True      REGULAR      USD  Call   TSLA  
1        True      REGULAR      USD  Call   TSLA  
2        True      REGULAR      USD  Call   TSLA  
3        True      REGULAR      USD  Call   TSLA  
4        True      REGULAR      USD  Call   TSLA  
5        True      REGULAR      USD  Call   TSLA  
6        True      REGULAR      USD  Call   TSLA  
7        True      REGULAR      USD  Call   TSLA  
8        True      REGULAR      USD  Call   TSLA  
9        True      REGULAR      USD  Call   TSLA

In [87]:
options_15sep23.isna().sum()

contractSymbol       0
lastTradeDate        0
strike               0
lastPrice            0
bid                  0
ask                  0
change               0
percentChange        0
volume               0
openInterest         0
impliedVolatility    0
inTheMoney           0
contractSize         0
currency             0
Type                 0
Ticker               0
dtype: int64

In [88]:
options_15sep23.rename(columns={'Ticker': 'company_id'}, inplace=True)

In [89]:
options_15sep23 = options_15sep23[['company_id'] + [col for col in options_15sep23.columns if col != 'company_id']]

In [90]:
options_15sep23 

company_id       contractSymbol             lastTradeDate  strike  \
0          TSLA  TSLA230915C00001670 2023-08-24 16:38:35+00:00    1.67   
1          TSLA  TSLA230915C00003330 2023-08-25 14:33:55+00:00    3.33   
2          TSLA  TSLA230915C00005000 2023-08-17 14:49:44+00:00    5.00   
3          TSLA  TSLA230915C00006670 2023-07-28 19:25:46+00:00    6.67   
4          TSLA  TSLA230915C00008330 2023-07-28 19:25:46+00:00    8.33   
...         ...                  ...                       ...     ...   
1992       MSFT  MSFT230915P00495000 2023-08-23 14:06:36+00:00  495.00   
1993       MSFT  MSFT230915P00500000 2023-08-16 18:39:22+00:00  500.00   
1994       MSFT  MSFT230915P00505000 2023-08-16 18:39:22+00:00  505.00   
1995       MSFT  MSFT230915P00510000 2023-08-16 18:39:22+00:00  510.00   
1996       MSFT  MSFT230915P00520000 2023-07-25 15:15:29+00:00  520.00   

      lastPrice    bid     ask  change  percentChange  volume  openInterest  \
0        231.66    0.0    0.00     0.0            0.0     3.0           0.0   
1        231.10    0.0    0.00     0.0            0.0     5.0           0.0   
2        220.15    0.0    0.00     0.0            0.0     2.0           0.0   
3        258.73  229.3  233.75     0.0            0.0     3.0          29.0   
4        257.08  227.9  232.20     0.0            0.0     1.0          17.0   
...         ...    ...     ...     ...            ...     ...           ...   
1992     168.32    0.0    0.00     0.0            0.0     4.0           0.0   
1993     176.60    0.0    0.00     0.0            0.0   145.0           0.0   
1994     181.50    0.0    0.00     0.0            0.0     0.0           0.0   
1995     186.15    0.0    0.00     0.0            0.0     0.0           0.0   
1996     172.75  195.6  196.45     0.0            0.0     0.0           0.0   

      impliedVolatility  inTheMoney contractSize currency  Type  
0              0.000010        True      REGULAR      USD  Call  
1              0.000010        True      REGULAR      USD  Call  
2              0.000010        True      REGULAR      USD  Call  
3             10.593753        True      REGULAR      USD  Call  
4              9.832035        True      REGULAR      USD  Call  
...                 ...         ...          ...      ...   ...  
1992           0.000010        True      REGULAR      USD   Put  
1993           0.000010        True      REGULAR      USD   Put  
1994           0.000010        True      REGULAR      USD   Put  
1995           0.000010        True      REGULAR      USD   Put  
1996           0.851564        True      REGULAR      USD   Put  

[1997 rows x 16 columns]

In [91]:
#options_15sep23.to_sql('option_info_15sep23', engine, if_exists='append', index=False)
#once pushed, I note it

3. Joining the expiration date and the option contracts

In [92]:
options_concat_col = options_15sep23[["company_id", "contractSymbol", "strike", "Type"]]

In [93]:
options_concat_col

company_id       contractSymbol  strike  Type
0          TSLA  TSLA230915C00001670    1.67  Call
1          TSLA  TSLA230915C00003330    3.33  Call
2          TSLA  TSLA230915C00005000    5.00  Call
3          TSLA  TSLA230915C00006670    6.67  Call
4          TSLA  TSLA230915C00008330    8.33  Call
...         ...                  ...     ...   ...
1992       MSFT  MSFT230915P00495000  495.00   Put
1993       MSFT  MSFT230915P00500000  500.00   Put
1994       MSFT  MSFT230915P00505000  505.00   Put
1995       MSFT  MSFT230915P00510000  510.00   Put
1996       MSFT  MSFT230915P00520000  520.00   Put

[1997 rows x 4 columns]

In [94]:
expr_dt

expiration_date
0        2013-01-18
1        2013-02-15
2        2013-03-15
3        2013-04-19
4        2013-05-17
..              ...
123      2023-04-21
124      2023-05-19
125      2023-06-16
126      2023-07-21
127      2023-08-18

[128 rows x 1 columns]

In [95]:
repeated_information = pd.concat([options_concat_col]*len(expr_dt), ignore_index=True)
repeated_information

company_id       contractSymbol  strike  Type
0            TSLA  TSLA230915C00001670    1.67  Call
1            TSLA  TSLA230915C00003330    3.33  Call
2            TSLA  TSLA230915C00005000    5.00  Call
3            TSLA  TSLA230915C00006670    6.67  Call
4            TSLA  TSLA230915C00008330    8.33  Call
...           ...                  ...     ...   ...
255611       MSFT  MSFT230915P00495000  495.00   Put
255612       MSFT  MSFT230915P00500000  500.00   Put
255613       MSFT  MSFT230915P00505000  505.00   Put
255614       MSFT  MSFT230915P00510000  510.00   Put
255615       MSFT  MSFT230915P00520000  520.00   Put

[255616 rows x 4 columns]

In [96]:
repeated_dates = pd.concat([expr_dt]*len(options_concat_col), ignore_index=True).sort_values(by='expiration_date', ascending=True).reset_index(drop=True)
repeated_dates

expiration_date
0           2013-01-18
1           2013-01-18
2           2013-01-18
3           2013-01-18
4           2013-01-18
...                ...
255611      2023-08-18
255612      2023-08-18
255613      2023-08-18
255614      2023-08-18
255615      2023-08-18

[255616 rows x 1 columns]

In [97]:
option_complete = pd.concat([repeated_dates, repeated_information], axis=1)
option_complete 

expiration_date company_id       contractSymbol  strike  Type
0           2013-01-18       TSLA  TSLA230915C00001670    1.67  Call
1           2013-01-18       TSLA  TSLA230915C00003330    3.33  Call
2           2013-01-18       TSLA  TSLA230915C00005000    5.00  Call
3           2013-01-18       TSLA  TSLA230915C00006670    6.67  Call
4           2013-01-18       TSLA  TSLA230915C00008330    8.33  Call
...                ...        ...                  ...     ...   ...
255611      2023-08-18       MSFT  MSFT230915P00495000  495.00   Put
255612      2023-08-18       MSFT  MSFT230915P00500000  500.00   Put
255613      2023-08-18       MSFT  MSFT230915P00505000  505.00   Put
255614      2023-08-18       MSFT  MSFT230915P00510000  510.00   Put
255615      2023-08-18       MSFT  MSFT230915P00520000  520.00   Put

[255616 rows x 5 columns]

In [98]:
checking = option_complete.loc[[1996, 1997]]
checking

expiration_date company_id       contractSymbol  strike  Type
1996      2013-01-18       MSFT  MSFT230915P00520000  520.00   Put
1997      2013-02-15       TSLA  TSLA230915C00001670    1.67  Call

4. Bringing the stock price in the moment of expiration and deciding if it's in the money

In [99]:
connection_string = 'mysql+pymysql://root:'+password+'@localhost/raw_data'
engine = create_engine(connection_string)
query = '''select si.company_id, si.date, si.close FROM stock_information as si;'''

stock_info = pd.read_sql_query(query, engine)

In [100]:
stock_info

company_id        date       close
0           TSLA  2013-01-02    2.357333
1           TSLA  2013-01-03    2.318000
2           TSLA  2013-01-04    2.293333
3           TSLA  2013-01-07    2.289333
4           TSLA  2013-01-08    2.245333
...          ...         ...         ...
17328       MSFT  2023-08-21  321.880005
17329       MSFT  2023-08-22  322.459991
17330       MSFT  2023-08-23  327.000000
17331       MSFT  2023-08-24  319.970001
17332       MSFT  2023-08-25  322.980011

[17333 rows x 3 columns]

In [109]:
expr_dt['expiration_date'] = pd.to_datetime(expr_dt['expiration_date'])
stock_info['date'] = pd.to_datetime(stock_info['date'])

In [110]:
merged_data = pd.merge(expr_dt, stock_info, left_on='expiration_date', right_on='date')
merged_data = merged_data.drop(columns=['date'])

In [113]:
merged_data

expiration_date company_id       close
0        2013-01-18       TSLA    2.301333
1        2013-01-18       META   29.660000
2        2013-01-18       GOOG   17.546980
3        2013-01-18       AAPL   17.857143
4        2013-01-18       AMZN   13.606000
..              ...        ...         ...
803      2023-08-18       GOOG  128.110001
804      2023-08-18        NIO   10.670000
805      2023-08-18       AAPL  174.490005
806      2023-08-18       AMZN  133.220001
807      2023-08-18       MSFT  316.480011

[808 rows x 3 columns]

In [114]:
merged_data.shape

(808, 3)

In [115]:
grouped_dataframes = {}

for company_id, company_group in merged_data.groupby('company_id'):
    grouped_dataframes[company_id] = company_group.copy()

In [126]:
tsla_dataframe = grouped_dataframes['TSLA']
tsla_dataframe.shape

(125, 3)

In [125]:
meta_dataframe = grouped_dataframes['TSLA']
meta_dataframe.shape

(125, 3)

In [124]:
nio_dataframe = grouped_dataframes['NIO']
nio_dataframe.shape

(58, 3)

In [ ]:
#['TSLA', 'META', 'GOOG', 'NIO', 'AAPL', 'AMZN', 'MSFT']

In [127]:
google_dataframe = grouped_dataframes['GOOG']
google_dataframe.shape

(125, 3)

In [129]:
amazon_dataframe = grouped_dataframes['AMZN']
amazon_dataframe.shape

(125, 3)

In [130]:
microsoft_dataframe = grouped_dataframes['MSFT']
microsoft_dataframe.shape

(125, 3)

We can see that we dont have 128 dates price for stock, but 125 for US companies, and NIO only has 58, as the company started in 2018. Once we link the stock price with each company_id and expiration date, we have to eliminate the Nan values.

In [131]:
merged_data

expiration_date company_id       close
0        2013-01-18       TSLA    2.301333
1        2013-01-18       META   29.660000
2        2013-01-18       GOOG   17.546980
3        2013-01-18       AAPL   17.857143
4        2013-01-18       AMZN   13.606000
..              ...        ...         ...
803      2023-08-18       GOOG  128.110001
804      2023-08-18        NIO   10.670000
805      2023-08-18       AAPL  174.490005
806      2023-08-18       AMZN  133.220001
807      2023-08-18       MSFT  316.480011

[808 rows x 3 columns]

In [132]:
option_complete 

expiration_date company_id       contractSymbol  strike  Type
0           2013-01-18       TSLA  TSLA230915C00001670    1.67  Call
1           2013-01-18       TSLA  TSLA230915C00003330    3.33  Call
2           2013-01-18       TSLA  TSLA230915C00005000    5.00  Call
3           2013-01-18       TSLA  TSLA230915C00006670    6.67  Call
4           2013-01-18       TSLA  TSLA230915C00008330    8.33  Call
...                ...        ...                  ...     ...   ...
255611      2023-08-18       MSFT  MSFT230915P00495000  495.00   Put
255612      2023-08-18       MSFT  MSFT230915P00500000  500.00   Put
255613      2023-08-18       MSFT  MSFT230915P00505000  505.00   Put
255614      2023-08-18       MSFT  MSFT230915P00510000  510.00   Put
255615      2023-08-18       MSFT  MSFT230915P00520000  520.00   Put

[255616 rows x 5 columns]

In [135]:
options_stock = option_complete.merge(merged_data, on=['expiration_date', 'company_id'], how='left')
options_stock

expiration_date company_id       contractSymbol  strike  Type  \
0           2013-01-18       TSLA  TSLA230915C00001670    1.67  Call   
1           2013-01-18       TSLA  TSLA230915C00003330    3.33  Call   
2           2013-01-18       TSLA  TSLA230915C00005000    5.00  Call   
3           2013-01-18       TSLA  TSLA230915C00006670    6.67  Call   
4           2013-01-18       TSLA  TSLA230915C00008330    8.33  Call   
...                ...        ...                  ...     ...   ...   
255611      2023-08-18       MSFT  MSFT230915P00495000  495.00   Put   
255612      2023-08-18       MSFT  MSFT230915P00500000  500.00   Put   
255613      2023-08-18       MSFT  MSFT230915P00505000  505.00   Put   
255614      2023-08-18       MSFT  MSFT230915P00510000  510.00   Put   
255615      2023-08-18       MSFT  MSFT230915P00520000  520.00   Put   

             close  
0         2.301333  
1         2.301333  
2         2.301333  
3         2.301333  
4         2.301333  
...            ...  
255611  316.480011  
255612  316.480011  
255613  316.480011  
255614  316.480011  
255615  316.480011  

[255616 rows x 6 columns]

In [137]:
options_stock.isna().sum()

expiration_date        0
company_id             0
contractSymbol         0
strike                 0
Type                   0
close              10413
dtype: int64

As said before, we can drop this Nan columns, as we are creating "artificially" the options

In [140]:
clean_options_stock = options_stock.dropna()

In [142]:
clean_options_stock.rename(columns={"close":"stock_price"}, inplace=True)

/var/folders/dm/hhj01rls7tz3g5v4m3hhj1bw0000gn/T/ipykernel_82331/1248489980.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_options_stock.rename(columns={"close":"stock_price"}, inplace=True)


In [143]:
clean_options_stock

expiration_date company_id       contractSymbol  strike  Type  \
0           2013-01-18       TSLA  TSLA230915C00001670    1.67  Call   
1           2013-01-18       TSLA  TSLA230915C00003330    3.33  Call   
2           2013-01-18       TSLA  TSLA230915C00005000    5.00  Call   
3           2013-01-18       TSLA  TSLA230915C00006670    6.67  Call   
4           2013-01-18       TSLA  TSLA230915C00008330    8.33  Call   
...                ...        ...                  ...     ...   ...   
255611      2023-08-18       MSFT  MSFT230915P00495000  495.00   Put   
255612      2023-08-18       MSFT  MSFT230915P00500000  500.00   Put   
255613      2023-08-18       MSFT  MSFT230915P00505000  505.00   Put   
255614      2023-08-18       MSFT  MSFT230915P00510000  510.00   Put   
255615      2023-08-18       MSFT  MSFT230915P00520000  520.00   Put   

        stock_price  
0          2.301333  
1          2.301333  
2          2.301333  
3          2.301333  
4          2.301333  
...             ...  
255611   316.480011  
255612   316.480011  
255613   316.480011  
255614   316.480011  
255615   316.480011  

[245203 rows x 6 columns]

Final step before running analysis and try to predict the volume --> check if it the option was or not in the money 

Call in the money stricking price < stock_price

Put in the money stricking price > stock_price

In [144]:
def calculate_in_the_money(row):
    if row['Type'] == 'Call' and row['stock_price'] > row['strike']:
        return 'Yes'
    elif row['Type'] == 'Put' and row['stock_price'] < row['strike']:
        return 'Yes'
    else:
        return 'No'

In [145]:
clean_options_stock['in_the_money'] = clean_options_stock.apply(calculate_in_the_money, axis=1)


/var/folders/dm/hhj01rls7tz3g5v4m3hhj1bw0000gn/T/ipykernel_82331/4128262031.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_options_stock['in_the_money'] = clean_options_stock.apply(calculate_in_the_money, axis=1)


In [156]:
clean_options_stock

expiration_date company_id       contractSymbol  strike  Type  \
0           2013-01-18       TSLA  TSLA230915C00001670    1.67  Call   
1           2013-01-18       TSLA  TSLA230915C00003330    3.33  Call   
2           2013-01-18       TSLA  TSLA230915C00005000    5.00  Call   
3           2013-01-18       TSLA  TSLA230915C00006670    6.67  Call   
4           2013-01-18       TSLA  TSLA230915C00008330    8.33  Call   
...                ...        ...                  ...     ...   ...   
255611      2023-08-18       MSFT  MSFT230915P00495000  495.00   Put   
255612      2023-08-18       MSFT  MSFT230915P00500000  500.00   Put   
255613      2023-08-18       MSFT  MSFT230915P00505000  505.00   Put   
255614      2023-08-18       MSFT  MSFT230915P00510000  510.00   Put   
255615      2023-08-18       MSFT  MSFT230915P00520000  520.00   Put   

        stock_price in_the_money  
0          2.301333          Yes  
1          2.301333           No  
2          2.301333           No  
3          2.301333           No  
4          2.301333           No  
...             ...          ...  
255611   316.480011          Yes  
255612   316.480011          Yes  
255613   316.480011          Yes  
255614   316.480011          Yes  
255615   316.480011          Yes  

[245203 rows x 7 columns]

Get the expiration dates of the clean_option_stock. This dataframe will rule the stock_filtered SQL

In [150]:
unique_exp = clean_options_stock["expiration_date"].unique()
unique_exp

array(['2013-01-18T00:00:00.000000000', '2013-02-15T00:00:00.000000000',
       '2013-03-15T00:00:00.000000000', '2013-04-19T00:00:00.000000000',
       '2013-05-17T00:00:00.000000000', '2013-06-21T00:00:00.000000000',
       '2013-07-19T00:00:00.000000000', '2013-08-16T00:00:00.000000000',
       '2013-09-20T00:00:00.000000000', '2013-10-18T00:00:00.000000000',
       '2013-11-15T00:00:00.000000000', '2013-12-20T00:00:00.000000000',
       '2014-01-17T00:00:00.000000000', '2014-02-21T00:00:00.000000000',
       '2014-03-21T00:00:00.000000000', '2014-05-16T00:00:00.000000000',
       '2014-06-20T00:00:00.000000000', '2014-07-18T00:00:00.000000000',
       '2014-08-15T00:00:00.000000000', '2014-09-19T00:00:00.000000000',
       '2014-10-17T00:00:00.000000000', '2014-11-21T00:00:00.000000000',
       '2014-12-19T00:00:00.000000000', '2015-01-16T00:00:00.000000000',
       '2015-02-20T00:00:00.000000000', '2015-03-20T00:00:00.000000000',
       '2015-04-17T00:00:00.000000000', '2015-05-15

In [154]:
expiration_dates_KEY = pd.DataFrame({'expiration_date': unique_exp})

In [155]:
expiration_dates_KEY

expiration_date
0        2013-01-18
1        2013-02-15
2        2013-03-15
3        2013-04-19
4        2013-05-17
..              ...
120      2023-04-21
121      2023-05-19
122      2023-06-16
123      2023-07-21
124      2023-08-18

[125 rows x 1 columns]

Sending the option dataframe to SQL and downloading the expiration_date_key dataframe

In [157]:
#clean_options_stock.to_sql('options_raw_information_v0', engine, if_exists='append', index=False)

245203

In [ ]:
expiration_dates_KEY.to_csv("expiration_date_KEY.csv", index=False) #exportar CSV.